### ISSUE TITLE
Clarification of alignment requirement for matrix columns

### ISSUE THREAD

**Issue by javagl** [KhronosGroup/glTF#2117]
The [specifcation, 3.6.2.4. Data Alignment](https://www.khronos.org/registry/glTF/specs/2.0/glTF-2.0.html#data-alignment) says:

> When two or more vertex attribute accessors use the same bufferView, its byteStride MUST be defined.
> ...
> For performance and compatibility reasons, each element of a vertex attribute MUST be aligned to 4-byte boundaries [...]
> 
> Accessors of matrix type have data stored in column-major order; start of each column MUST be aligned to 4-byte boundaries. [...]

The first statements clearly refer to _vertex attributes_. And knowing about the care and diligence that went into all this, it's reasonable to assume that these requirements do _not_ apply to data that is _not_ used as vertex attributes (otherwise, the statements would not be restricted to these).  

The last statement does **not** explicitly refer to vertex attributes. It seems to talk about accessors _in general_. 

So the question is:

Does the alignment requirement for matrix columns apply to **ALL** accessors, or **ONLY** to vertex attributes?

If it referred only to vertex attributes, the statement should be changed to

> Accessors of matrix type have data stored in column-major order. **When they are used as vertex attributes, then the** start of each column MUST be aligned to 4-byte boundaries. [...]

If it referred to all accessors, it might make sense to say that explicitly, to avoid confusion. Maybe by just adding something like "**(regardless of whether they are used as vertex attributes or not)**" as a clarification, unless someone comes up with a more elegant way of phrasing that...

**Comment by lexaknyazev** [comment]
Matrix accessors are barely used (besides 4x4 skinning matrices) and thus may not even be fully implemented in real engines. There are several important implementation details here and one genuine spec issue.

GPUs don't support matrix type vertex attributes directly so each matrix column effectively uses a separate vertex attribute slot and should therefore be treated as a separate attribute. Since regular vertex attributes require the 4-byte alignment, it has to apply to matrix columns too. Consider this an externally-imposed requirement.

From the glTF point of view, an element of a `MAT*` accessor is the whole matrix and therefore `bufferView.byteStride` means the distance between the first elements of two matrices, not the distance between columns of one matrix. Therefore, per-column alignment cannot be expressed with glTF properties as-is. For this reason, the spec _always_ requires extra padding for 2x2 and 3x3 matrices of certain component types to enforce this.

Let's take a `MAT2` accessor with 8-bit values for example. The generic 4-byte alignment rule (checked with `accessor.byteOffset`) can only enforce the alignment of the first column of the first matrix so two extra bytes of padding are needed to ensure that the second column is aligned too. Since accessors are supposed to fully define how data is read from a buffer view, those per-column padding bytes must always exist, regardless of whether the accessor is used for vertex attributes or not (usage may define compatibility of the accessor object but not its layout).

The open question that's worth rectifying here is the total byte size of a matrix accessor element when such padding is present. Continuing with a `MAT2` accessor using 8-bit values:
```
 00  01  02  03  04  05  06  07
m00 m10  XX  XX m01 m11  ??  ??
```

If this matrix accessor is used for a vertex attribute, it _has to_ align the start of each column of each matrix to 4 bytes, therefore the buffer view must set its `byteStride` property to 8 (at least) to align the start of the second matrix.

However, what should happen if that accessor is used for something other than a vertex attribute?

The spec explicitly forbids setting the `bufferView.byteStride` if that buffer view is not used for vertex attributes. Therefore, one could argue that the effective size of such accessor elements is 6 bytes. Following the same logic, the effective element size of `MAT3` accessors using 8-bit values would be 11 bytes and the effective element size of `MAT3` accessors using 16-bit values would be 22 bytes.

The spec should either make these quite unusual numbers official, or it should say that these three matrix configurations always include per-column trailing bytes (thus enforcing accessor element sizes of 8, 12, and 24 respectively).

Note that this issue only affects three aforementioned matrix configurations that are not even used by the core spec or any well-known extensions.

/cc @bghgary @emackey @aaronfranke 

---

The most plausible use case not related to vertex attributes would be a JSON property that represents a 2x2 or 3x3 matrix (as a four- or nine-element array respectively) that is animated via the `KHR_animation_pointer` extension. In that case, an implementation would have to read multiple matrix values from a buffer view with no explicit stride and that requires knowing correct offsets.

**Comment by javagl** [comment]
Zooming into that specific detail: I wonder about the significance of the 'effective size'. Is this supposed to be come a precisely defined, normative term?

To my understanding, this is _only_ relevant for the narrow case where someone has an accessor that contains these matrices, and _explicitly_ wants to omit the last 1 or 2 bytes from the underlying buffer view, exploiting the statement

> Alignment requirements apply only to the start of each column, so trailing bytes MAY be omitted if there’s no further data.

from the bottom of https://registry.khronos.org/glTF/specs/2.0/glTF-2.0.html#data-alignment

On the one hand, I see the point that this is "externally-imposed". On the other hand, every library that wants to handle `MATn` types programmatically (in the case that they are _not_ vertex attributes) will have to deal with the resulting quirks. One of these quirks is that the size of a MAT2/byte accessor is **twice** the size that it would have to be. Another one is that the elements of such an accessor can not "trivially" be interpreted as matrices.

---

A specific example is https://github.com/CesiumGS/cesium-native/issues/677 : The "goal" there is/was, roughly speaking, to have something like a `Matrix` class that represents the MAT2/byte matrix as contiguous memory, and doing something like
```
const std::byte* accesorData = accessor.getData();
const Matrix* matrices =  reinterpret_cast<const Matrix*>(accessorData);

const Matrix matrix = matrices[0]; // This is not a valid matrix!!!
```

---

I'd rather lean towards tweaking the _"... so trailing bytes MAY be omitted"_ statement for MATn accessors that are not vertex attributes, but that's just a gut feeling. Let's see what others think about that...

**Comment by lexaknyazev** [comment]
> I wonder about the significance of the 'effective size'. Is this supposed to be come a precisely defined, normative term?

Yes, because when the `byteStride` is not defined, the elements are "tightly" packed and that packing directly depends on the "effective size".

Ultimately, this issue comes down to the following question: what is the byte offset of the second accessor element of an accessor with `MAT2` type and 8-bit component type that uses a buffer view without defined `byteStride`. Is it 6 or 8? And if it's 6, does it implicitly make the accessor inherently invalid if its `count` is greater than one (because the first column of the second matrix wouldn't be aligned)?

---

The statement about the trailing bytes of the last accessor element generally refers to the case when the byte stride is greater than the accessor element size and thus such bytes are not accessed. For example, a `VEC3` accessor with an 8-bit component type that has two elements and uses a buffer view with `byteStride` of 4 would only need to access 7 bytes from the buffer view. Since the `byteStride` property cannot be used with non-vertex data, this statement is not supposed to be applicable in non-vertex cases.

**Comment by javagl** [comment]
It may not be _supposed_ to be applicable, but as it stands, it is independent of the byte stride. At least, I don't see a reason why it shouldn't cover an accessor with MAT2/byte values
`m00 m10  XX  XX m01 m11` 
that refers to a 6-byte buffer view.
(Maybe it could be clarified, but the exact wording then likely depends on how _this_ issue is supposed to be resolved).

---

I might not have all implications on the radar for the 6-vs-8 byte option. So just to be sure: The 6-byte version would indeed mean that accessors that are **not** vertex attributes are packed as
`m00 m10  XX  XX m01 m11 m00 m10  XX  XX m01 m11`
But if they **are** vertex attributes, they have to be packed using the 8-byte pattern:
`m00 m10  XX  XX m01 m11 XX XX m00 m10  XX  XX m01 m11 XX XX` (or
`m00 m10  XX  XX m01 m11 XX XX m00 m10  XX  XX m01 m11`, omitting the last padding bytes)


If this is correct (and the only relevant difference), then I'd lean towards the 8-byte layout. Whether or not accessor data is used as a vertex attribute (either now, via an extension, or in the future) is something that should _preferably_ not affect the data layout. Yes, it already does, via the padding. The point is that even though this padding is odd and wasteful, we cannot get rid of it anyhow, so I think that it should be the same for vertex attributes and non-vertex-attributes. Having different data layouts will cause subtle bugs and inconsistencies. (Or to argue with pseudocode: It turns a `memcpy` into a `for` loop...)

**Comment by lexaknyazev** [comment]
The implication is correct but the description is a bit inaccurate. The generic alignment requirement for vertex attributes already enforces the 8-byte layout (at least) and that works with the current spec language just by setting the byte stride correctly. Note that the byte stride may be set to values larger than 8. So let's ignore the vertex attribute case for now.

If the accessor is not used for vertex attributes, it _cannot_ use a buffer view with a byte stride. Therefore, its elements are tightly packed by definition, i.e., there's no trailing bytes (logically). This means that 6 or 8 bytes is the intrinsic byte length of a single MAT2/8-bit element, i.e., all padding bytes exist "inside" the element and therefore the accessor always takes `(6 or 8) * count` bytes. Thus this part resolves the "last trailing bytes" question by bypassing it completely.

Overall, I'm leaning towards specifying that a single MAT2/8-bit accessor element always takes 8 bytes (with similar statements about MAT3/8-bit and MAT3/16-bit taking 12 and 24 bytes respectively).
